cpp/src/arrow/io/memory.cc

```diff
@@ -199,11 +199,11 @@ class FixedSizeBufferWriter::FixedSizeBufferWriterImpl {
   int64_t memcopy_threshold_;
 };
 
-FixedSizeBufferWriter::~FixedSizeBufferWriter() {}
-
 FixedSizeBufferWriter::FixedSizeBufferWriter(const std::shared_ptr<Buffer>& buffer)
     : impl_(new FixedSizeBufferWriterImpl(buffer)) {}
 
+FixedSizeBufferWriter::~FixedSizeBufferWriter() = default;
```

Comment:
This is new syntax on me, is there a reason to prefer this over the empty dtor?

Reply:
Sorry to answer a quesion with another question, but do we actually need any of the destructors for the classes that are not holding a `std::unique_ptr`? They are all trivial destructors and can be generated by the compiler.

Reply:
I don’t know the answer to this, it would be good to have a definitive answer. Definitely if you do not have a virtual destructor in the base class, you will leak memory / fail to destruct members of subclasses

Reply:
Right, that makes sense. When you say `default` it forces the destructor to be compiler generated, according to http://en.cppreference.com/w/cpp/language/destructor.

In the same reference it also says this about implicitly declared destructors:

> If no user-declared destructor is provided for a class type (struct, class, or union), the compiler will always declare a destructor as an inline public member of its class.

Googling around, [this SO thread](https://stackoverflow.com/q/17541671/564538) suggests that trivial destructors don't need to be declared or defined (modulo the base class issue).

In the first reference there's also this statement:
> For both user-defined or implicitly-defined destructors, after the body of the destructor is executed, the compiler calls the destructors for all non-static non-variant members of the class, in reverse order of declaration, then it calls the destructors of all direct non-virtual base classes in reverse order of construction (which in turn call the destructors of their members and their base classes, etc), and then, if this object is of most-derived class, it calls the destructors of all virtual bases.

Coupled with the compiler-generated definition being an empty body and the fact that the destruction path is the same for user-defined and implicitly defined destructors, I don't think we need to declare or define these constructors (modulo the base class issue + classes holding `std::unique_ptr`s.